### PR TITLE
Make OutputType consistent with Descriptor and return it

### PR DESCRIPTION
**Pull request by Sjors**
This PR renames `OutputType::LEGACY` to `OutputType::PKH`, `OutputType::P2SH_SEGWIT` to `OutputType::P2SH_WPKH` and `OutputType::BECH32` to `OutputType::WPKH` in order to be consistent with the terminology used in [Output Descriptors](https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md) and in order to more accurately describe what the current wallet uses.

It introduces two new OutputTypes `P2SH`, `P2SH_WSH` and `WSH` to distinguish legacy P2SH from SegWit wrapped P2SH, and to distinguish SegWit Pay-to-Public-Key from Pay-to-Script-Hash.

Finally it adds `GetAddressType()` to `Descriptor`, which returns the reintroduced (#12729) `OutputType::NONE` for descriptors that we currently can't convert to an address (`pk(...)`, `raw()`, etc), those whose address we don't interpret (`addr()`) and the quirky `combo()`.

This allows a future descriptor based wallet (e.g. #15487) to find the correct descriptor for a given `-addresstype` and `-changetype` setting when generating new addresses. Similarly it allows importing the appropriate descriptor from a hardware wallet (e.g. #14912).

### CODE REVIEW

**Comment by MeshCollider** [comment]
Concept ACK, will take a look :)

**Comment by ryanofsky** [comment]
It would be better to use Optional<OutputType> than to add an OutputType::NONE enum. There may be some contexts where it's valid for the output type to be unset, but in most places it's not valid, and using the optional type would do a better job communicating and enforcing this.

File: src/outputtype.cpp [hunk 1]
```diff
@@ -20,13 +20,13 @@ static const std::string OUTPUT_TYPE_STRING_BECH32 = "bech32";
 bool ParseOutputType(const std::string& type, OutputType& output_type)
 {
     if (type == OUTPUT_TYPE_STRING_LEGACY) {
-        output_type = OutputType::LEGACY;
+        output_type = OutputType::PKH;
```
Comment (MarcoFalke) [inline src/outputtype.cpp]: style-nit: in commit b2da4c74d891fe719b21411f692e3edd4b968c

Do as scripted diff?

**Comment by sipa** [comment]
I don't understand the motivation for this.

When you request a Bech32 address from the wallet, getting a P2WPKH or P2WSH are both fine, and which one depends on the context currently (getnewaddress vs addmultisigaddress). In a future descriptor-based design there should just be one default source for all Bech32 addresses - which can be either P2WPKH or P2WSH (which depends on the setup and spending policy you want, not on what you're requesting.

**Comment by Sjors** [comment]
@ryanofsky I also thought about using `Optional`. I can switch to that unless others see a problem.

@sipa we current use OutputType for three different tasks

1. distinguish between P2PKH, P2SH-WPKH and P2WSH. That distinction is historical, because it happens to be the three different output types the wallet supports. Most other (non-multisig) wallets also use these three types.
2. distinguish what we expect other wallets can handle. All existing wallets can send to both P2PKH and P2SH-WPKH, but not all wallets can send to P2WSH.
3. keep support for generating fresh P2PKH addresses

I think we should introduce a new type for task (2): `AddressType`: bech32 or base58. Then there could be a simple mapping from OutputType to AddressType. For backwards compatibility all we need for OutputType is P2PKH, P2SH-WPKH and P2WSH, so I could remove the others, although they're trivial.

As you pointed out, for descriptor based wallets, we only need to care about `AddressType`. The wallet can link to one base58 and one bech32 descriptor. Those descriptors would be set upon wallet creation, by default they would be P2SH-WPKH (bech32) and P2WSH (base58). Someone can also create a legacy wallet with only P2PKH (base58) and no bech32.   (this assumes that we encourage users to create multiple wallets if they want both P2PKH and P2SH-WPKH)

For a descriptor wallet that relies on external keys, e.g. BIP44/49/84, we could fetch the BIP49 (base58) and BIP84 (bech32) derivation xpubs from the device.

**Comment by sipa** [comment]
@Sjors Not really. LEGACY refers to pkh or p2sh-multisig; P2SH_SEGWIT refers to p2sh-pwpkh or p2sh-p2wsh-multisig; BECH32 refers to p2wpkh or p2wsh-multisig.

They indeed serve two different goals (a) selecting between bech32 support in the receiver or not and (b) choosing non-segwit outputs.

I still don't see what in de codebase needs further differentiation between multisig or not etc.

**Comment by Sjors** [comment]
IIUC the only interaction the user has with `OutputType` is through `getnewaddress` and `addmultisigaddress`. So far I've only considered `getnewaddress`, where `legacy` refers to `pkh`.

With `addmultisigaddress` the `address_type` argument would map to different output types, i.e. `P2SH`, `P2SH_WSH` and `WSH`. That's still compatible with the idea of a simple `OutputType` -> `AddressType` mapping in descriptor wallets.

I'm not advocating `OutputType` to be even more fine grained. E.g. `MultisigDescriptor` would have `OutputType:: P2SH_WSH`.


An alternative could be to "deprecate" `OutputType`, i.e. only use it in the context of a legacy wallet.  The `bitcoind -address_type` parameter could be ignored for descriptor wallets, and the `address_type` argument for  `getnewaddress` could be constrained to `bech32` and `base58` for descriptor wallets. (`addmultisigaddress` is not needed for descriptor wallets)

In that case this PR won't be necessary, but then I do have to find another solution for deciding which descriptor to import from a hardware wallet into a legacy wallet (#14912). Right now it considers the current `address_type` and then matches the descriptor string prefix. However it might be better to make that PR contingent on descriptor wallets anyway, because if the user switch `-addresstype` or calls `getnewaddress` with a different address type than during key import, they'll end up generating addresses on the wrong derivation path. That breaks compatibility with the manufacturer software wallets which assume BIP44/49/84. Descriptor wallets solve that problem, because each descriptor can have its own derivation path.

**Comment by Sjors** [comment]
I'm trying a different approach in #15590.

**Pull request closed without merging**